Split API info into another zip entry

eed3si9n opened the pull request:
It's been reported that for large code bases with thousands of files,
the Analysis object reaches the size of 100MB, mostly from the API info
blob.

This changes SameAPI.apply to use only the API hash, and thereby
removing the necessity to load previous API info during the incremental
compilation.

To prevent needless reading of API info, the companions map is now
stored into another zip entry, which is read lazily.

/cc @gkossakowski, @romanowski

smarter commented on internal/zinc-core/src/main/scala/sbt/internal/inc/IncrementalCommon.scala:
```diff
@@ -85,25 +85,22 @@ private[inc] abstract class IncrementalCommon(log: sbt.util.Logger, options: Inc
-          val apiUnifiedPatch = apiDiff.generateApiDiff(src.toString, oldApi.api, newApi.api, contextSize)
-          wrappedLog.debug(s"Detected a change in a public API ($src):\n$apiUnifiedPatch")
+          wrappedLog.debug(s"""Detected a change in a public API ($src):
+                              |old hash = ${oldApi.apiHash}, new hash = ${newApi.apiHash}""".stripMargin)
```
Wait, does that mean that API diffs are no longer printed? They are _extremely_ useful when debugging/developing incremental compilation

eed3si9n replied:
The previous API info is still around, just not loaded in during the compilation, so we might be able to do API diffing after the compilation is done.

gkossakowski commented on internal/compiler-interface/src/main/datatype/other.json:
```diff
@@ -16,7 +16,7 @@
 				{
 					"name": "api",
-					"type": "Companions"
+					"type": "xsbti.Maybe<Companions>"
```
Why it becomes optional now?

eed3si9n replied:
When I save to `inc_compile` I throw this away, and it thaws back as `Maybe.nothing()` value here, which should help in both time and space.

gkossakowski commented:
As far as I understand, this change moves away zinc from operating on api objects. It switches solely to api hashes. It's pretty change big so the commit message should be expanded accordingly. The questions I think commit message should answer are:
- what size of the project makes API info file so large that it's a significant performance hit to load it?
- what about removed API diffs? API diffs are the key identifying tricky problems in incremental compiler, e.g.: https://github.com/sbt/zinc/issues/88 (and http://www.scala-sbt.org/0.13/docs/Understanding-Recompilation.html#Debugging+an+interface+representation)

eed3si9n commented:
@gkossakowski Replaced the commit to based on the review.
The subprojects I'm looking at contains 1000+ source files, which then could further expand via code generation, macro, compiler plugin etc. Also it's not uncommon for Play projects to contain dozens of subprojects, which on their own could expand as teams grow.

I've brought back the API diffing, so all you have to do is enable the flag and load the full API info, which can be done via `getWithAPI` on `AnalysisStore` vs plain `get`.

smarter commented on internal/zinc-core/src/main/scala/sbt/internal/inc/IncrementalCommon.scala:
```diff
@@ -94,16 +95,24 @@ private[inc] abstract class IncrementalCommon(log: sbt.util.Logger, options: Inc
-          wrappedLog.debug(s"Detected a change in a public API ($src):\n$apiUnifiedPatch")
+          (m2o(oldApi.api), m2o(newApi.api)) match {
+            // The API info now is optional. The caller is responsible for loading the previous Analysis
+            // with the full API info using the info stored in an APIInfoStore.
```
I think this should be done automatically all the time when debug is enabled

gkossakowski commented:
I did a simple test of compiling an empty class to see how much space we spend on API representation. The number I got is roughly 3.8kb. To get to 100mb of API representation you need around 25k top level classes. The API representation compresses very well using standard zip compression. I get 13.5x contraction in size.

Having said that, splitting API into a separate storage is fine. I think @smarter's latest feedback is spot on and should be addressed.

eed3si9n commented:
@gkossakowski @smarter Moved the "with-or-without API info" logic into the `AnalysisStore`.

romanowski commented:
I agree with @gkossakowski  about simple zip compression - in our we get around 10% of file caches. See my PR: #160

IMO correct way to split APIs to another files is to extract them from AnalyzedClass to separate lazy entry (map) in Analysis itself.

eed3si9n commented:
ok. Turned it into lazy instead of optional as per @romanowski's suggestion:

``` diff
                {
                    "name": "api",
 -                  "type": "Companions"
 +                  "type": "lazy Companions"
                },
```

This does make the whole thing more "automatic" as read-API-or-not-to-read decision is deferred to the usage point.

romanowski commented on internal/zinc-core/src/main/scala/sbt/internal/inc/AnalysisStore.scala:
```diff
@@ -6,9 +6,11 @@ package internal
 
 trait AnalysisStore {
   def set(analysis: CompileAnalysis, setup: MiniSetup): Unit
+  /** Calls `getWithAPI` if API debug is enabled, calls `getWithoutAPI` otherwise. */
```
This comment is outdated

romanowski commented on internal/zinc-core/src/main/scala/sbt/internal/inc/IncrementalNameHashing.scala:
```diff
@@ -21,7 +21,7 @@ private final class IncrementalNameHashing(log: sbt.util.Logger, options: IncOpt
     apis: APIs): Set[String] = {
     transitiveDeps(invalidatedClasses, logging = false)(relations.inheritance.internal.reverse) filter {
-      className => APIUtil.hasPackageObject(apis.internalAPI(className))
+      _.endsWith(".package") // APIUtil.hasPackageObject(apis.internalAPI(className))
```
Please remove commented code

romanowski commented on internal/zinc-persist/src/main/scala/sbt/internal/inc/FileBasedStore.scala:
```diff
@@ -7,17 +7,53 @@ package inc
     def getUncaught(): (CompileAnalysis, MiniSetup) =
-      Using.fileReader(IO.utf8)(file) { reader => TextAnalysisFormat.read(reader) }
+      {
+        val (analysis0: Analysis, setup) = Using.fileReader(IO.utf8)(analysisFile) { reader => TextAnalysisFormat.read(reader) }
```
I think all this logic should be moved to  APIsF's read and write methods (in TextAnalysisFormat.scala) not to construct twice AnalyzedClass instance.

romanowski commented on internal/zinc-persist/src/main/scala/sbt/internal/inc/FileBasedStore.scala:
```diff
@@ -7,17 +7,53 @@ package inc
+      }
+  }
+
+  private final class FileBasedAPIInfoStore0(file: File) {
```
Why 0?

romanowski commented on internal/zinc-persist/src/main/scala/sbt/internal/inc/FileBasedStore.scala:
```diff
@@ -7,17 +7,53 @@ package inc
+    new FileBasedStore0(analysisFile, apiInfoFile)
+
+  private final class FileBasedStore0(analysisFile: File, apiInfoFile: File) extends AnalysisStore {
+    val apiInfoStore = new FileBasedAPIInfoStore0(apiInfoFile)
```
I think this is great candidate for 2nd entry in zip... ( #160 )

romanowski commented on internal/zinc-persist/src/main/scala/sbt/internal/inc/TextAnalysisFormat.scala:
```diff
@@ -266,6 +279,28 @@ object TextAnalysisFormat {
     }
   }
 
+  private[this] object FullAPIsF {
```
I think we should store there only companions - we don't want to store compilations and name hashes twice (especially name hashes are heavy)

romanowski commented on internal/zinc-persist/src/main/scala/xsbt/api/AnalyzedClassFormat.scala:
```diff
@@ -3,20 +3,34 @@
-  }
+object AnalyzedClassFormats {
+  // This will throw out API information intentionally.
+  def lightFormat(implicit ev0: Format[Compilation], ev1: Format[NameHashes]): Format[AnalyzedClass] =
```
If we store just companions in 2nd store then we can use single format

romanowski commented on internal/zinc-persist/src/main/scala/xsbt/api/CompanionsFormat.scala:
```diff
@@ -0,0 +1,16 @@
+import xsbti.api._
+import sbinary._
+
+object CompanionsFormat extends Format[Companions] {
```
If we move all compatnions to single store then we can serialized whole map not single companion one by one. I think we should let Java optimized our serialization

eed3si9n replied:
I want to rely less on Java serialization.. eventually I want to turn all these into JSON so we can cross machine boundaries safely.

eed3si9n committed:
Split API info into another zip entry

It's been reported that for large code bases with thousands of files,
the Analysis object reaches the size of 100MB, mostly from the API info
blob.

This changes SameAPI.apply to use only the API hash, and thereby
removing the necessity to load previous API info during the incremental
compilation.

To prevent needless reading of API info, the companions map is now
stored into another zip entry, which is read lazily.

eed3si9n commented:
@romanowski Thanks for the detailed review. I did adopt most of your suggestions, and pushing the responsibility down to TextAnalysisFormat did make the imple much cleaner. I also made the Companions storage another zip entry.

romanowski commented on internal/zinc-persist/src/main/scala/sbt/internal/inc/FileBasedStore.scala:
```diff
@@ -6,48 +6,67 @@ package internal
-    private val entryFileName = "cache.txt"
+  def apply(file: File): AnalysisStore = new FileBasedStore0(file)
+
+  private final class FileBasedStore0(file: File) extends AnalysisStore {
```
maybe `ZippedStore`?

eed3si9n replied:
The whole point is that I'm mimicking the `object FileBaseStore` so I think that's fine. It's only so the stack trace looks nicer.

romanowski replied:
I don't want to nitpick but I just don't like 0s in class names :)

eed3si9n replied:
How about `Impl`?

romanowski replied:
:+1:

romanowski commented on internal/zinc-persist/src/main/scala/sbt/internal/inc/FileBasedStore.scala:
```diff
@@ -6,48 +6,67 @@ package internal
+          val internal = apis.internal map { case (k, v) => k -> v.api }
+          val external = apis.external map { case (k, v) => k -> v.api }
+          outputStream.putNextEntry(new ZipEntry(companionsFileName))
+          TextAnalysisFormat.writeCompanionMap(writer, internal, external)
```
can we pass whole APIs class here and extract this at the last possible place (CompanionsF IIRC)?

It should make code nicer.

eed3si9n replied:
Sure.

romanowski commented:
Except minor comments above LGTM

eed3si9n committed:
Some changes per review
